crypto-ffi/bindings/js/CoreCrypto.ts

```diff
@@ -20,6 +20,81 @@ import wasm from "../../../crypto-ffi/Cargo.toml";
 
 import type * as CoreCryptoFfiTypes from "./wasm/core-crypto-ffi";
 
+interface CoreCryptoRichError {
+    errorName: string;
+    message: string;
+    rustStackTrace: string;
+    proteusErrorCode: number;
+}
+
+/**
+ * Error wrapper that takes care of extracting rich error details across the FFI (through JSON parsing)
+ *
+ * Whenever you're supposed to get this class (that extends `Error`) you might end up with a base `Error`
+ * in case the parsing of the message structure fails. This is unlikely but the case is still covered and fall backs automatically.
+ * More information will be found in the base `Error.cause` to inform you why the parsing has failed.
+ *
+ * Please note that in this case the extra properties will not be available.
+ */
+export class CoreCryptoError extends Error {
+    rustStackTrace: string;
+    proteusErrorCode: number;
+
+    private constructor(msg: string, richError: CoreCryptoRichError, ...params: any[]) {
+        // @ts-ignore
+        super(msg, ...params);
+        Object.setPrototypeOf(this, new.target.prototype);
+
+        this.name = richError.errorName;
+        this.rustStackTrace = richError.rustStackTrace;
+        this.proteusErrorCode = richError.proteusErrorCode;
+    }
+
+    private static fallback(msg: string, ...params: any[]): Error {
+        console.warn("Cannot build CoreCryptoError, falling back to standard Error");
+        // @ts-ignore
+        return new Error(msg, ...params);
+    }
+
+    static build(msg: string, ...params: any[]): CoreCryptoError|Error {
+        const parts = msg.split("\n\n");
+        if (parts.length < 2) {
+            const cause = new Error("CoreCrypto WASM FFI Error doesn't have enough elements to build a rich error");
+            return this.fallback(msg, { cause }, ...params);
+        }
+
+        const [errMsg, richErrorJSON, ] = parts;
+        try {
+            const richError: CoreCryptoRichError = JSON.parse(richErrorJSON);
+            return new this(errMsg, richError, ...params);
+        } catch(cause) {
+            return this.fallback(msg, { cause }, ...params);
+        }
+    }
+
+    static fromStdError(e: Error): CoreCryptoError|Error {
+        const opts = {
+            // @ts-ignore
+            cause: e.cause || undefined,
+            stack: e.stack || undefined,
+        };
+
+        return this.build(e.message, opts);
+    }
+
+    static async asyncMapErr<T>(p: Promise<T>): Promise<T> {
+        const mappedErrorPromise = p.catch((e: Error|CoreCryptoError) => {
+            if (e instanceof CoreCryptoError) {
+                throw e;
+            } else {
+                throw this.fromStdError(e);
+            }
+        });
+
+        return await mappedErrorPromise;
+    }
```

Comment:
Not important since this seems to be hidden from the callers, but I'd remove this from this class. When its being called I have an impression that we're always returning an error but in fact we're returning a "Result" that is mapping the possible error to this.

Reply:
I isolated this into a function because of the sheer amount of times this code is called (51 to be accurate). And it didn't make sense to me to put it anywhere else than as a static method on the error class

JS `Promise` is basically a Rust `async` `Result<T, E>` anyway; you either resolve the Promise successfully - with a value - or you throw within the promise - with usually an instance of `Error` - So it made sense to me to make it work very much like Rust's `.map_err()`

Reply:
I agree. My comment is in regards of calling it: you call `CoreCryptoError. asyncMapErr` which leaves an impression that an error will always be returned but in fact it's still the Promise/Result. My comment was more to extract it from the `CoreCryptoError` and put it in something called `CoreCrypto` or `CoreCryptoResult`. As I said, not important as our callers don't see this, so feel free to ignore